log: fix race in TestLogEntryPropagation

abarganier opened the pull request:
Fixes: https://github.com/cockroachdb/cockroach/issues/122791

TestLogEntryPropagation was unsafely reassigning `OrigStderr` in the log
package which was tripping the race detector.

This patch makes use of the existing `redirectStderr` function which
allows us to safely reassign stderr.

Release note: none

xinhaoz commented on pkg/util/log/stderr_redirect.go:
```diff
@@ -12,26 +12,86 @@ package log
+}
+
+// File returns a reference to the underlying file. Generally, this is
+// only useful if you're about to call SwapFile, and you want to hold onto
```
Why not make `SwapFile` return the old file? Then we don't need this fn.

xinhaoz reviewed (commented):
Are we unable to use https://github.com/cockroachdb/cockroach/blob/2f9f3d686e76cf3a106dd4cf5e70c48fc60ef97b/pkg/util/log/stderr_redirect.go#L47 or export this fn, https://github.com/cockroachdb/cockroach/blob/2f9f3d686e76cf3a106dd4cf5e70c48fc60ef97b/pkg/util/log/stderr_redirect.go#L41 in the tests to accomplish this?

Looks like there are thread safe utils in the log pkg for this already, but not sure if we want to export them.

abarganier reviewed (commented):
> Are we unable to use

I was looking at this mutex and `hijackStderr` as well, but I believe their purpose is to safely reassign `os.Stderr`, whereas the work done here is to ensure proper reassignment of `log.OrigStderr`. LMK if I'm missing something - this `hijackStderr` stuff isn't something I've dealt with prior to this PR. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/124219#-:-NyQcSvh58aTM84wGJf6:b-d12rq7)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @kyle-a-wong and @xinhaoz)

___
*[`pkg/util/log/stderr_redirect.go` line 70 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/124219#gh-1607347379:-NyQcRfN92czqtgOtmh6:b1iebdp) ([raw file](https://github.com/cockroachdb/cockroach/blob/8e510b5f2c6eefdd766843d5e516f6d65bcd0334/pkg/util/log/stderr_redirect.go#L70)):*
<details><summary><i>Previously, xinhaoz (Xin Hao Zhang) wrote…</i></summary><blockquote>

Why not make `SwapFile` return the old file? Then we don't need this fn.
</blockquote></details>

Good idea, done. 


<!-- Sent from Reviewable.io -->

abarganier reviewed (commented):
Hold off on reviewing, I'm going to take another look based on our discussion at standup today.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/124219#-:-NyQojbN0rJLFRUD8zQR:b-q2wuk6)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @kyle-a-wong and @xinhaoz)



<!-- Sent from Reviewable.io -->

abarganier committed:
log: fix race in TestLogEntryPropagation

Fixes: https://github.com/cockroachdb/cockroach/issues/122791

TestLogEntryPropagation was unsafely reassigning `OrigStderr` in the log
package which was tripping the race detector.

This patch makes use of the existing `hijackStderr` function which
allows us to safely reassign stderr.

Release note: none

xinhaoz reviewed (approved):
Nice! Thank you for the changes 👍

dhartunian commented:
bors r=xinhaoz
